In [1]:
#importing packages
import imutils
import cv2

In [2]:
#images are rerpresented as multi-dimensional NumPy array with
#dimensions height* columns

image = cv2.imread("test.jpeg")
image

array([[[207, 208, 198],
        [207, 208, 198],
        [208, 209, 199],
        ...,
        [106, 127, 125],
        [106, 127, 125],
        [105, 126, 124]],

       [[206, 207, 197],
        [207, 208, 198],
        [207, 208, 198],
        ...,
        [106, 127, 125],
        [105, 126, 124],
        [105, 126, 124]],

       [[206, 207, 197],
        [206, 207, 197],
        [207, 208, 198],
        ...,
        [105, 126, 124],
        [105, 126, 124],
        [105, 126, 124]],

       ...,

       [[218, 209, 195],
        [218, 209, 195],
        [221, 210, 196],
        ...,
        [ 46,  87, 110],
        [ 47,  85, 109],
        [ 46,  84, 108]],

       [[218, 209, 195],
        [218, 209, 195],
        [221, 210, 196],
        ...,
        [ 44,  85, 108],
        [ 45,  83, 107],
        [ 45,  83, 107]],

       [[220, 209, 195],
        [220, 209, 195],
        [220, 209, 195],
        ...,
        [ 42,  83, 106],
        [ 41,  82, 105],
        [ 41,  82, 105]]

In [3]:
(h,w,d) = image.shape
#to extract the height width and depth
print("width={}, height={}, depth={}".format(w,h,d))

width=640, height=443, depth=3


In [4]:
#displaying the image
cv2.imshow("Image",image)
cv2.waitKey(0)
cv2.destroyAllWindows()

<font color = "#757494"><H3>Till here we loaded and displayed the picture</H3></font>

<font color = "#757494"><H1>WHAT IS A PIXEL?</h1></font>
<p>Images are made of pixels in a grid. A 640 x 480 image has 640 columns (the width) and 480 rows (the height). There are 640 * 480 = 307200  pixels in an image with those dimensions.</p>
<p>Each pixel in a grayscale image has a value representing the shade of gray. In OpenCV, there are 256 shades of gray — from 0 to 255. So a grayscale image would have a grayscale value associated with each pixel.</p>
<p>Pixels in a color image have additional information. There are several color spaces that you’ll soon become familiar with as you learn about image processing. For simplicity let’s only consider the RGB color space.
    
In OpenCV color images in the RGB (Red, Green, Blue) color space have a 3-tuple associated with each pixel: (B, G, R) .</p>

In [5]:
#accessing the RGB pixel located at x=50, y=100, NOTE - REMEMBER OPEN CV TAKES BGR instead of RGB
(B,G,R) = image[100,50] #here on this line we have taken 'y' before the 'x' value as images are stored in the form
#of height(rows)xwidth(cols) i.e. Y*X!
print("R={}, G={}, B={}".format(R,G,B))

R=213, G=216, B=209


In [6]:
#array slicing and cropping i.e. extracting regions of intrest(ROI)
#let's extract 100*100 pixel square ROI from the input image
roi = image[60:160,320:420]  #image[startY:endY,startX:endX]
cv2.imshow("Region of Intrest", roi)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [7]:
#now resizing image, not taking aspect ratio into count
resized = cv2.resize(image,(200,200))
cv2.imshow("Here's the resized image",resized)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [8]:
#resizing image taking aspect ratio into count
#taking width as 300 and it will automatically give the aspect ratio
r = 300.0/w  #
dim = (300, int(h*r))
resized = cv2.resize(image,dim)
cv2.imshow("resized with ratio in mind",resized)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [9]:
#for that we have imutils function already defined.
resized = imutils.resize(image,width=300)
cv2.imshow("resized image",resized)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [10]:
#rotating the image by 45 degrees
#for that we need to compute the center of the image
#then we will have to construct the rotation matrix
#then finally we will apply affine warp

center =(w//2,h//2)
M = cv2.getRotationMatrix2D(center,-45,1.0)
rotated = cv2.warpAffine(image,M,(w,h))
cv2.imshow("Rotated Image",rotated)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [11]:
#same can be done with imutils library
rotated = imutils.rotate(image, -45)
cv2.imshow("Rotated Image",rotated)
cv2.waitKey(0)
cv2.destroyAllWindows()

<h4>Here on rotating we found that our image was clipped</h4>
<p>So we will use an imutils function in order to overcome that problem</p>

In [12]:
#rotatinf without the image clipping
rotated = imutils.rotate_bound(image,-45)
cv2.imshow("Rotated Image",rotated)
cv2.waitKey(0)
cv2.destroyAllWindows()

<h3>Image smoothing</h3>
<p>In many image processing pipelines, we must blur an image to reduce high-frequency noise, making it easier for our algorithms to detect and understand the actual contents of the image rather than just noise that will “confuse” our algorithms.

Refer here for more on blurrness - https://www.pyimagesearch.com/2016/07/25/convolutions-with-opencv-and-python/
</p>


In [13]:
blurred = cv2.GaussianBlur(image,(11,11),0)
cv2.imshow("Bluured Image",blurred)
cv2.waitKey(0)
cv2.destroyAllWindows()

<h3>Drawing on an image</h3>
<p>Before we move on with drawing on an image with OpenCV, take note that drawing operations on images are performed in-place. Therefore at the beginning of each code block, we make a copy of the original image storing the copy as output . We then proceed to draw on the image called output in-place so we do not destroy our original image.</p>

In [14]:
#here we are drawing a 2px thick rectangle
output = image.copy()
cv2.rectangle(output,(320,60),(420,160),(0,0,255),2) #topleft coordinate and then bottom right
cv2.imshow("Rectangle",output)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [15]:
#drawing a solid circle at x = 300, y=150
output = image.copy()
cv2.circle(output,(300,150),20,(255,0,0),-1) #(imagesrc,location,thickness,color,-1 denotes solid)
cv2.imshow("Circle",output)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [16]:
#here we are drawing a 5px thick line
output = image.copy()
output = image.copy()
cv2.line(output, (60, 20), (400, 200), (0, 0, 255), 5)
cv2.imshow("Line", output)
cv2.waitKey(0)
cv2.destroyAllWindows()

In [ ]:
# draw green text on the image
output = image.copy()
cv2.putText(output, "LEARNING OPEN CV BITCH", (10, 25), 
	cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0,0,255), 2)
cv2.imshow("Text", output)
cv2.waitKey(0)
cv2.destroyAllWindows()